File: ios/yifan/YifanJustifiedTextManager.m
#import <React/RCTViewManager.h>

#import "YifanJustifiedTextView.h"

@interface YifanJustifiedTextManager : RCTViewManager
@end

@implementation YifanJustifiedTextManager

RCT_EXPORT_MODULE(YifanJustifiedText)

- (UIView *)view {
  return [[YifanJustifiedTextView alloc] init];
}

RCT_EXPORT_VIEW_PROPERTY(segments, NSArray)
RCT_EXPORT_VIEW_PROPERTY(textColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(accentColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(tagActiveColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(tagActiveBackgroundColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(tagInactiveBackgroundColor, UIColor)
RCT_EXPORT_VIEW_PROPERTY(fontFamily, NSString)
RCT_EXPORT_VIEW_PROPERTY(fontSize, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(lineHeight, CGFloat)
RCT_EXPORT_VIEW_PROPERTY(justify, BOOL)
RCT_EXPORT_VIEW_PROPERTY(activeTag, NSString)

RCT_EXPORT_VIEW_PROPERTY(onPressMention, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPressTag, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPressLink, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onPressText, RCTBubblingEventBlock)
RCT_EXPORT_VIEW_PROPERTY(onContentSizeChange, RCTBubblingEventBlock)

@end

File: android/app/src/main/java/im/cxa/fanatter/YifanJustifiedTextManager.kt
package im.cxa.fanatter

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.SimpleViewManager
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp

class YifanJustifiedTextManager : SimpleViewManager<YifanJustifiedTextView>() {

  override fun getName(): String = NAME

  override fun createViewInstance(reactContext: ThemedReactContext): YifanJustifiedTextView {
    return YifanJustifiedTextView(reactContext)
  }

  @ReactProp(name = "segments")
  fun setSegments(view: YifanJustifiedTextView, segments: ReadableArray?) {
    val list = mutableListOf<Map<String, Any?>>()
    if (segments != null) {
      for (i in 0 until segments.size()) {
        val map = segments.getMap(i) ?: continue
        list.add(map.toHashMap())
      }
    }
    view.segments = list
  }

  @ReactProp(name = "textColor", customType = "Color")
  fun setTextColor(view: YifanJustifiedTextView, color: Int?) {
    if (color != null) view.textColorValue = color
  }

  @ReactProp(name = "accentColor", customType = "Color")
  fun setAccentColor(view: YifanJustifiedTextView, color: Int?) {
    if (color != null) view.accentColorValue = color
  }

  @ReactProp(name = "tagActiveColor", customType = "Color")
  fun setTagActiveColor(view: YifanJustifiedTextView, color: Int?) {
    if (color != null) view.tagActiveTextColor = color
  }

  @ReactProp(name = "tagActiveBackgroundColor", customType = "Color")
  fun setTagActiveBackgroundColor(view: YifanJustifiedTextView, color: Int?) {
    view.tagActiveBackground = color
  }

  @ReactProp(name = "tagInactiveBackgroundColor", customType = "Color")
  fun setTagInactiveBackgroundColor(view: YifanJustifiedTextView, color: Int?) {
    view.tagInactiveBackground = color
  }

  @ReactProp(name = "fontFamily")
  fun setFontFamily(view: YifanJustifiedTextView, family: String?) {
    view.customFontFamily = family
  }

  @ReactProp(name = "fontSize", defaultFloat = 15f)
  fun setFontSize(view: YifanJustifiedTextView, size: Float) {
    view.fontSizeDp = if (size > 0f) size else 15f
  }

  @ReactProp(name = "lineHeight", defaultFloat = 24f)
  fun setLineHeight(view: YifanJustifiedTextView, height: Float) {
    view.lineHeightDp = if (height > 0f) height else 24f
  }

  @ReactProp(name = "justify", defaultBoolean = true)
  fun setJustify(view: YifanJustifiedTextView, justify: Boolean) {
    view.justifyEnabled = justify
  }

  @ReactProp(name = "activeTag")
  fun setActiveTag(view: YifanJustifiedTextView, tag: String?) {
    view.activeTag = tag
  }

  override fun getExportedCustomDirectEventTypeConstants(): Map<String, Any> =
      mapOf(
          "topPressMention" to mapOf("registrationName" to "onPressMention"),
          "topPressTag" to mapOf("registrationName" to "onPressTag"),
          "topPressLink" to mapOf("registrationName" to "onPressLink"),
          "topPressText" to mapOf("registrationName" to "onPressText"),
          "topContentSizeChange" to mapOf("registrationName" to "onContentSizeChange"),
      )

  companion object {
    const val NAME = "YifanJustifiedText"
  }
}
